src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -820,20 +841,22 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
         }
 
         private static MetadataBinder BuildNewMetadataBinder(ODataUriParserConfiguration config,
+            IEdmNavigationSource navigationSource,
             IEdmNavigationSource targetNavigationSource,
             IEdmTypeReference elementType,
             HashSet<EndPathToken> generatedProperties = null,
             bool collapsed = false)
         {
-            BindingState state = CreateBindingState(config, targetNavigationSource, elementType, generatedProperties, collapsed);
+            BindingState state = CreateBindingState(config, navigationSource, targetNavigationSource, elementType, generatedProperties, collapsed);
             return new MetadataBinder(state);
         }
 
         private static BindingState CreateBindingState(ODataUriParserConfiguration config,
-            IEdmNavigationSource targetNavigationSource, IEdmTypeReference elementType,
-            HashSet<EndPathToken> generatedProperties = null, bool collapsed = false)
+            IEdmNavigationSource navigationSource, IEdmNavigationSource targetNavigationSource,
+            IEdmTypeReference elementType, HashSet<EndPathToken> generatedProperties = null,
+            bool collapsed = false)
         {
-            if (targetNavigationSource == null && elementType == null)
+            if (navigationSource == null && targetNavigationSource == null && elementType == null)

```

Comment:
I think adding `navigationSource == null` to this condition makes it possible to get a null-reference exception online 868.

With the original if statement condition `(targetNavigationSource == null && elementType == null)`   if `elementType` is null, then we can guarantee that `targetNavigationSource` is non-null on line 845, otherwise the method would have terminated on line 861.

But with the new condition, it's possible that both `targetNavigationSource` and `elementType` are `null` on line 868 (if `navigationSource` is non-null), leading to a potential null-reference exception on `targetNavigationSource.EntityType()`

Reply:
#Resolved

Reply:
Good catch!

---
In reply to: [530665718](https://github.com/OData/odata.net/pull/1906#discussion_r530665718) [](ancestors = 530665718)